jsonrpc/ipc.go

```diff
@@ -0,0 +1,221 @@
+package jsonrpc
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/p2p/netutil"
+	"github.com/sourcegraph/conc"
+)
+
+type Ipc struct {
+	endpoint string
+	rpc      *Server
+	events   NewRequestListener
+	log      utils.SimpleLogger
+
+	connWg sync.WaitGroup // connWg is a WaitGroup for tracking active connections.
+
+	connParams IpcConnParams
+	listener   net.Listener
+
+	socketCallback func() // socketCallback is executed when socket has been created
+
+	// everything below is protected
+	mu      sync.Mutex
+	conns   map[net.Conn]struct{} // conns is a map that holds active connections.
+	closing bool                  // closing indicates whether the server is in the process of shutting down.
+}
+
+// NewIpc creates a new IPC handler instance with the provided RPC server, endpoint and logger.
+func NewIpc(rpc *Server, endpoint string, log utils.SimpleLogger) *Ipc {
+	return &Ipc{
+		endpoint:       endpoint,
+		rpc:            rpc,
+		log:            log,
+		connParams:     DefaultIpcConnParams(),
+		conns:          make(map[net.Conn]struct{}),
+		events:         &SelectiveListener{},
+		socketCallback: func() {},
+	}
+}
+
+// WithSocketCallback sets the callback for socket creation.
+func (i *Ipc) WithSocketCallback(f func()) *Ipc {
```

Comment:
Is this mainly for avoiding race conditions in tests?

Reply:
Yes. It used to be at initialization of handler instead of startup but I changed it in order to follow the http and ws services pattern at `node/http`.

Reply:
What do you think about replacing `NewIpc` with `DialIPC`, which does everything `NewIpc` does plus setting up the listener? That way we know the listener is initialized when we call `Run`.

Reply:
I don't mind reverting that change. As I said I changed it just to allign more or less with the `node/http.go` . (They construct listeneres at startup and initialization of struct never fails)
Regarding the DialIPC I think that it doesn't reflect the behaviour and can be too confusing. The DialIPC might even be required to implement the part you said above about checking if socket is alive

Reply:
Ah my bad, I misunderstood what you meant by setting up the listener "at the initialization of the handler." In this case, I think we can disregard the convention and initialize the listener in `NewIpc`. I think it's ok if `NewIpc` can return an error.

(apologies if my earlier statement that we should "maintain consistency with websocket and http handlers, and spawn a socket for each server" was was confusing here)

> Regarding the DialIPC I think that it doesn't reflect the behaviour and can be too confusing. 

Makes sense.

> The DialIPC might even be required to implement the part you said above about checking if socket is alive

Mind expanding on this? Feel free to move this part of the conversation to the other comment as well.

Reply:
Nevermind it, I thought that some simple ack might be required to check for listener but the dial is actually able to determine listener presence.